File: data/modules/texting/texting.py
"""Texting module — backend.

Self-contained drop-in. Registers a single prompt-block filter that
drops engine blocks tagged ``environmental`` when the focal character
is responding to a user message marked as a text.

The module never writes new prompt blocks of its own — its whole
backend job is to GATE existing engine blocks. Each environmental
block has its own per-setting strip toggle so a user can keep some
context (e.g., keep wardrobe but drop surroundings) without forking
the prompt assembler.

The frontend half (`texting.js`) stamps
``metadata.modules.texting = {to: <char_id>}`` on the user message
before POST; this filter reads that metadata at prompt-assembly time
to decide whether to drop blocks.
"""
from __future__ import annotations

from typing import Any

from app.modules.api import is_active as _is_active, settings_for as _settings_for
from app.prompt import register_filter, register, Block, register_message_annotator


MODULE_ID = "texting"


def _settings(conversation: dict[str, Any], leaf_id: str | None = None) -> dict[str, Any]:
    """Effective settings for this branch — defaults filled in.

    Mirrors what other modules do: read module_settings off the
    active setup root and fall back to manifest defaults for any
    keys the user hasn't customized. Hardcoded fallback values
    match data/modules/texting/module.json — if you change the
    manifest defaults, change them here too.
    """
    raw = _settings_for(conversation, MODULE_ID, leaf_id) or {}
    return {
        "strip_surroundings":     raw.get("strip_surroundings", True),
        "strip_others_present":   raw.get("strip_others_present", True),
        "strip_items_in_scene":   raw.get("strip_items_in_scene", True),
        "strip_scene_effects":    raw.get("strip_scene_effects", True),
        "strip_wardrobe_overrides": raw.get("strip_wardrobe_overrides", True),
        "render_style":           raw.get("render_style", "sms_bubble"),
    }


# Map between engine block id and the corresponding "strip_X" setting.
# When the setting is True, the block is dropped for texting replies.
_BLOCK_TO_SETTING = {
    "surroundings":        "strip_surroundings",
    "others_present":      "strip_others_present",
    "items_in_scene":      "strip_items_in_scene",
    "scene_effects":       "strip_scene_effects",
    "wardrobe_overrides":  "strip_wardrobe_overrides",
}


def _focal_is_replying_to_text(ctx) -> bool:
    """Walk recent history leaf→root and return True iff the most
    recent USER message has ``metadata.modules.texting.to`` pointing
    at the focal character.

    The "most recent user message" is the last message authored by
    the user (persona=='user') before any character/narrator turns.
    History is already in chronological order (root→leaf) via
    `path_to_root`, so we iterate in reverse.

    Returns False when there's no focal, no history, no recent user
    message, or the recent user message isn't a text addressed to
    THIS focal.
    """
    focal_id = ctx.focal_id
    if not focal_id:
        return False
    for msg in reversed(ctx.history or []):
        if msg.get("persona") != "user":
            continue
        meta = (msg.get("metadata") or {}).get("modules") or {}
        texting_meta = meta.get(MODULE_ID) or {}
        target = texting_meta.get("to")
        return isinstance(target, str) and target == focal_id
    return False


@register_filter
def _texting_strip_environmental(entry, block, ctx) -> bool:
    """Drop environmental blocks when the focal is replying to a text.

    Filter returns False to drop. Runs for every block the assembler
    emits; fast-paths to True (keep) when texting isn't active for
    this conversation, when the block isn't environmental, or when
    the user message we're replying to wasn't addressed at the focal.
    """
    if "environmental" not in (entry.tags or ()):
        return True  # not our concern
    if not _is_active(ctx.conversation, MODULE_ID, ctx.leaf_id):
        return True  # module isn't on for this branch
    if not _focal_is_replying_to_text(ctx):
        return True  # not a text reply
    setting_key = _BLOCK_TO_SETTING.get(entry.id)
    if not setting_key:
        # Environmental block we don't have a per-setting toggle
        # for. Default to keeping it — author can add a toggle later
        # if they want the strip.
        return True
    settings = _settings(ctx.conversation, ctx.leaf_id)
    # Setting True = drop the block; False = keep.
    return not settings.get(setting_key, True)


@register(id="texting_phone_primer", order=55, applies_to=("character",))
def _texting_phone_primer(ctx):
    """Prime the reply, on a text turn, to open with the character
    pulling out / checking their phone — grounding the SMS exchange as
    a physical act rather than a disembodied chat line.

    Fires ONLY when the texting module is active and the focal is
    replying to a text addressed at them (same gate as the environmental
    strip), so ordinary in-person turns are untouched. Phrased to skip
    the re-pull mid-exchange ('unless she already has it in hand'), so a
    back-and-forth doesn't narrate her fishing the phone out every line.
    """
    if not _is_active(ctx.conversation, MODULE_ID, ctx.leaf_id):
        return None
    if not _focal_is_replying_to_text(ctx):
        return None
    from app.personas import apply_macros

    content = (
        "{{char}} is not in the room with {{user}} — this is a text message. "
        "{{char}} is wherever she happens to be, and her phone has just buzzed "
        "with {{user}}'s text. Open the reply with her noticing and pulling out "
        "/ picking up her phone and reading it — unless she already has it in "
        "hand from an ongoing back-and-forth — then thumbing out her response. "
        "Keep the *action* beats on her and her phone; she can't see {{user}}'s "
        "surroundings.\n"
        "\n"
        "Example:\n"
        "{{user}}: hey, you up?\n"
        "{{char}}: *her phone buzzes face-down on the nightstand; she reaches "
        "over and squints at the bright screen in the dark, thumbing out a "
        "reply* \"mm... barely. what's up?\""
    )
    return Block(
        label="Texting",
        content=apply_macros(content, ctx.macros),
        section="Texting",
    )


@register_message_annotator
def _annotate_text_reply(conversation, message):
    """Persist the SMS marker onto a character reply that's answering a
    text, so the texting render style re-applies on reload instead of
    being re-derived client-side from a fragile immediate-parent walk.

    A reply counts as a text iff the NEAREST user message in its
    ancestry is a text addressed at this reply's speaker — mirroring the
    prompt-side `_focal_is_replying_to_text`. The marker keys off that
    persisted user-message metadata, so it self-gates: no user-side text
    marker, no annotation. Branch-specific by construction — it lives on
    this node, on this path, exactly like `applied_edits`.
    """
    persona = message.get("persona")
    if persona in (None, "user", "narrator"):
        return None
    speaker_id = message.get("speaker_id") or persona
    if not isinstance(speaker_id, str):
        return None
    msgs = conversation.get("messages") or {}
    # Skip this message itself if it's somehow already marked.
    own = ((message.get("metadata") or {}).get("modules") or {}).get(MODULE_ID)
    if isinstance(own, dict) and own:
        return None
    cur = msgs.get(message.get("parent_id"))
    seen = 0
    while cur is not None and seen < 200:  # cycle/runaway guard
        seen += 1
        if cur.get("persona") == "user":
            tmeta = ((cur.get("metadata") or {}).get("modules") or {}).get(MODULE_ID) or {}
            target = tmeta.get("to")
            if isinstance(target, str) and target == speaker_id:
                return {"modules": {MODULE_ID: {"to": speaker_id, "reply": True}}}
            # Nearest user message isn't a text to me — not a text reply.
            return None
        cur = msgs.get(cur.get("parent_id"))
    return None

File: data/modules/texting/texting.js
/**
 * Texting module — frontend.
 *
 * Self-contained drop-in. Mounts a "Text" button in the composer-right
 * region (owned by this module, so the engine auto-removes it when the
 * module is toggled off — see Modules.mount({owner}) / _fireActivationChange),
 * picks a target character from the in-cast list, stamps
 * metadata.modules.texting = {to: <char_id>} on the next user message
 * via Modules.onCompose, and decorates messages with the configured
 * bubble style via Modules.onMessage.
 *
 * Loaded only when the module's .js file exists on disk
 * (data/modules/texting/texting.js). Activation is checked at init
 * via Modules.isActive("texting") — inactive = the module is inert,
 * no UI mounted, no compose / message hooks register.
 */
(function () {
  const MODULE_ID = "texting";

  if (!window.Modules) {
    console.warn("texting: window.Modules unavailable; module dormant");
    return;
  }
  if (!Modules.isActive(MODULE_ID)) {
    return;  // module not active for this branch
  }

  // Internal state. `currentTarget` is the character id the next user
  // message gets stamped against. Reset to null after each send so
  // texting is single-turn unless the user re-picks before the next
  // message.
  let currentTarget = null;

  // ---- Compose hook: stamp metadata on the outgoing user message ----
  Modules.onCompose((pending) => {
    if (pending.persona !== "user") return;
    if (!currentTarget) return;
    pending.metadata = pending.metadata || {};
    pending.metadata.modules = pending.metadata.modules || {};
    pending.metadata.modules[MODULE_ID] = { to: currentTarget };
    // Single-turn: clear the target after stamping so the next
    // message is a regular in-person turn unless the user clicks
    // Text again.
    currentTarget = null;
    _refreshButtonLabel();
  });

  // URL override for the render style — lets the user A/B
  // sms_bubble / tinted / inline_badge without re-staging. Visit
  // any chat page with ?texting_style=tinted (or ...=inline_badge,
  // or ...=sms_bubble) to override the saved setting for THIS
  // page load only. Leave the param off to use the staged setting.
  const _urlStyleOverride = (() => {
    try {
      const v = new URLSearchParams(location.search).get("texting_style");
      if (v === "sms_bubble" || v === "tinted" || v === "inline_badge") return v;
    } catch (_e) { /* no-op */ }
    return null;
  })();

  // ---- Message hook: apply bubble CSS to texting turns ----
  Modules.onMessage((messageEl, message) => {
    const settings = Modules.settings(MODULE_ID) || {};
    const style = _urlStyleOverride || settings.render_style || "sms_bubble";
    // Primary path: the message carries its own persisted texting
    // marker. The user's outgoing text has {to: <char>}; a reply that
    // was answering a text is stamped {to: <char>, reply: true} by the
    // backend message annotator — so both re-render correctly on reload
    // straight from branch-persisted node metadata, no derivation.
    const own = (message?.metadata?.modules || {})[MODULE_ID];
    if (own) {
      messageEl.classList.add("texting-message");
      messageEl.classList.add(`texting-style-${style}`);
      if (own.reply) {
        messageEl.classList.add("texting-reply");
      } else {
        messageEl.dataset.textingTo = own.to || "";
      }
      return;
    }
    // Fallback for replies persisted before the marker existed (or any
    // the annotator didn't reach): walk the FULL ancestry to the nearest
    // user message and inherit its marker. Full-ancestry (not just the
    // immediate parent) so multi-response chains and intervening
    // narrator / auto-state nodes still resolve.
    const msgs = (typeof state !== "undefined" && state) ? state.conversation?.messages : null;
    if (!msgs) return;
    let cur = msgs[message?.parent_id];
    let guard = 0;
    while (cur && guard++ < 200) {
      if (cur.persona === "user") {
        const pm = (cur.metadata?.modules || {})[MODULE_ID];
        const toMe = pm && (message?.speaker_id ? pm.to === message.speaker_id : true);
        if (toMe) {
          messageEl.classList.add("texting-message", "texting-reply");
          messageEl.classList.add(`texting-style-${style}`);
        }
        return;  // nearest user message decides; stop either way
      }
      cur = msgs[cur.parent_id];
    }
  });

  // ---- UI: Text button + target picker ----
  const wrap = document.createElement("span");
  wrap.className = "texting-composer-wrap";

  const btn = document.createElement("button");
  btn.type = "button";
  btn.className = "ghost";
  btn.dataset.textingBtn = "1";
  btn.title = "Compose a text to a specific character. Their reply " +
              "strips environmental prompt blocks (Surroundings, etc).";
  wrap.appendChild(btn);

  function _refreshButtonLabel() {
    if (!currentTarget) {
      btn.textContent = "Text";
      btn.classList.remove("texting-armed");
      return;
    }
    const ent = state?.entities?.[currentTarget];
    const name = (ent && ent.name) || currentTarget;
    btn.textContent = `→ ${name}`;
    btn.classList.add("texting-armed");
  }

  function _eligibleCharacters() {
    // In-cast NPCs from the effective_cast set, excluding user. The
    // engine's state.effectiveCastChars carries the live branch cast.
    const out = [];
    const cast = state?.effectiveCastChars;
    if (!cast) return out;
    for (const cid of cast) {
      if (cid === "user") continue;
      const e = state.entities && state.entities[cid];
      if (!e || e.type !== "character") continue;
      out.push({ id: cid, name: e.name || cid });
    }
    out.sort((a, b) => a.name.localeCompare(b.name));
    return out;
  }

  function _showPicker() {
    // Close any existing picker first.
    const old = document.querySelector(".texting-picker");
    if (old) { old.remove(); return; }

    const choices = _eligibleCharacters();
    if (!choices.length) {
      // Nothing to text. Surface gently — don't crash.
      btn.classList.add("texting-empty");
      setTimeout(() => btn.classList.remove("texting-empty"), 1200);
      return;
    }

    const menu = document.createElement("div");
    menu.className = "texting-picker";
    for (const c of choices) {
      const opt = document.createElement("button");
      opt.type = "button";
      opt.className = "ghost texting-picker-row";
      opt.textContent = c.name;
      opt.addEventListener("click", (e) => {
        e.preventDefault();
        currentTarget = c.id;
        _refreshButtonLabel();
        menu.remove();
      });
      menu.appendChild(opt);
    }
    // Cancel row clears the armed target.
    const cancel = document.createElement("button");
    cancel.type = "button";
    cancel.className = "ghost texting-picker-cancel";
    cancel.textContent = "Cancel";
    cancel.addEventListener("click", (e) => {
      e.preventDefault();
      currentTarget = null;
      _refreshButtonLabel();
      menu.remove();
    });
    menu.appendChild(cancel);

    wrap.appendChild(menu);
    // Close on outside-click (next tick so this click doesn't catch).
    setTimeout(() => {
      const off = (ev) => {
        if (!menu.contains(ev.target) && ev.target !== btn) {
          menu.remove();
          document.removeEventListener("click", off);
        }
      };
      document.addEventListener("click", off);
    }, 0);
  }

  btn.addEventListener("click", (e) => {
    e.preventDefault();
    if (currentTarget) {
      // Click an armed button to clear.
      currentTarget = null;
      _refreshButtonLabel();
      return;
    }
    _showPicker();
  });

  _refreshButtonLabel();
  // Owned mount: the engine tracks this element against the module id and
  // auto-unmounts it when texting is toggled off mid-conversation (fixes the
  // button lingering after deactivation). "composer-right" resolves to the
  // engine's [data-region] anchor; the legacy "composer_right" SLOTS name
  // would also work via region()'s fallback.
  Modules.mount("composer-right", wrap, { owner: MODULE_ID });
})();
